### PR TITLE
support IN and CONTAINS

### LINKED ISSUE #2054: Add IN queries support in Firestore
Firestore supports IN queries now, so we can add support for that in name-based queries.

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #2054

### CODE REVIEW

File: docs/src/main/asciidoc/firestore.adoc [hunk 1]
```diff
@@ -146,6 +146,13 @@ Cloud Firestore only supports filter components joined by AND, and the following
 * `less than`
 * `is null`
 
+Operations expecting List arguments:
+
+* `contains` (allows for combining up to 10 array-contains clauses on the same field with a logical OR)
+* `in` (allows for combining up to 10 equality clauses on the same field with a logical OR)
+
+NOTE: Only one of `contains` and `in` operations are allowed in a query.
```
Comment (dzou) [inline docs/src/main/asciidoc/firestore.adoc]: replace 'and' with 'or'

File: docs/src/main/asciidoc/firestore.adoc [hunk 2]
```diff
@@ -146,6 +146,13 @@ Cloud Firestore only supports filter components joined by AND, and the following
 * `less than`
 * `is null`
 
+Operations expecting List arguments:
```
Comment (meltsufin) [inline docs/src/main/asciidoc/firestore.adoc]:
```suggestion
Operations expecting `List` arguments:
```
Also, I probably would just join with the list above. I think people understand that `in` and `contains` requires lists.

File: docs/src/main/asciidoc/firestore.adoc [hunk 3]
```diff
@@ -146,6 +146,13 @@ Cloud Firestore only supports filter components joined by AND, and the following
...
+
+* `contains` (allows for combining up to 10 array-contains clauses on the same field with a logical OR)
```
Comment (meltsufin) [inline docs/src/main/asciidoc/firestore.adoc]: I'm confused by this "10" limit.
Can you just explain that limitations in the note below?

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java [hunk 4]
```diff
@@ -69,6 +71,8 @@
 					.put(GREATER_THAN, StructuredQuery.FieldFilter.Operator.GREATER_THAN)
 					.put(LESS_THAN_EQUAL, StructuredQuery.FieldFilter.Operator.LESS_THAN_OR_EQUAL)
 					.put(LESS_THAN, StructuredQuery.FieldFilter.Operator.LESS_THAN)
+					.put(IN, StructuredQuery.FieldFilter.Operator.IN)
+					.put(CONTAINING, StructuredQuery.FieldFilter.Operator.ARRAY_CONTAINS_ANY)
```
Comment (meltsufin) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: Firestore has both `contains` and `contains-any`. Why are you using `contains-any` here vs `contains`?
Also, how can we support both types of contains?

Comment (dmitry-s) [reply spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: `contains-any` with a one element list is exactly the same as `contains`, so I think we don't need to support `contains` separately

Comment (meltsufin) [reply spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: What about with a list of more than one element?

Comment (dmitry-s) [reply spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: `contains` supports only singular values and `contains-any` combines `contain` filters using `OR`

Comment (meltsufin) [reply spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: There's this caveat in the [documentation](https://cloud.google.com/firestore/docs/query-data/queries):

> You can combine `array-contains` with `in` but not with `array-contains-any`.

**Comment by meltsufin** [comment]
Ping.

File: docs/src/main/asciidoc/firestore.adoc (hunk 2 above)
Comment (meltsufin) [reply docs/src/main/asciidoc/firestore.adoc]: Can you apply this little fix?

File: docs/src/main/asciidoc/firestore.adoc (hunk 3 above)
Comment (meltsufin) [reply docs/src/main/asciidoc/firestore.adoc]: Can you please address?

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java [hunk 5]
```diff
@@ -64,13 +66,18 @@
...
 	private final FirestoreClassMapper classMapper;
 
-	private static final Map<Part.Type, StructuredQuery.FieldFilter.Operator> PART_TO_FILTER_OP =
-			new MapBuilder<Part.Type, StructuredQuery.FieldFilter.Operator>()
-					.put(SIMPLE_PROPERTY, StructuredQuery.FieldFilter.Operator.EQUAL)
-					.put(GREATER_THAN_EQUAL, StructuredQuery.FieldFilter.Operator.GREATER_THAN_OR_EQUAL)
-					.put(GREATER_THAN, StructuredQuery.FieldFilter.Operator.GREATER_THAN)
-					.put(LESS_THAN_EQUAL, StructuredQuery.FieldFilter.Operator.LESS_THAN_OR_EQUAL)
-					.put(LESS_THAN, StructuredQuery.FieldFilter.Operator.LESS_THAN)
+	private static final Map<Part.Type, OperatorSelector> PART_TO_FILTER_OP =
```
Comment (meltsufin) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: I think this should move to the `OperatorSelector` class for better encapsulation.

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java [hunk 6]
```diff
@@ -137,13 +144,14 @@ public Object execute(Object[] parameters) {
 					throw new FirestoreDataException(
 							"Too few parameters are provided for query method: " + getQueryMethod().getName());
 				}
-				StructuredQuery.FieldFilter.Operator filterOp = PART_TO_FILTER_OP.get(part.getType());
-				if (filterOp == null) {
+				Object value = it.next();
+				OperatorSelector operatorSelector = PART_TO_FILTER_OP.get(part.getType());
```
Comment (meltsufin) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: I would replace this access of the map with
```java
OperatorSelector.getFilter(part, value)
```

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java [hunk 7]
```diff
@@ -156,4 +164,25 @@ public Object execute(Object[] parameters) {
...
+			this.operatorForSingleType = operatorForSingleType;
+			this.operatorForIterableType = operatorForIterableType;
+		}
+
+		OperatorSelector(StructuredQuery.FieldFilter.Operator commonOperator) {
+			this.operatorForSingleType = commonOperator;
+			this.operatorForIterableType = commonOperator;
+		}
+
+		StructuredQuery.FieldFilter.Operator getFiler(Object val) {
```
Comment (elefeint) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/repository/query/PartTreeFirestoreQuery.java]: s/getFiler/getFilter

File: spring-cloud-gcp-data-firestore/src/test/java/org/springframework/cloud/gcp/data/firestore/it/FirestoreRepositoryIntegrationTests.java [hunk 8]
```diff
@@ -166,6 +168,73 @@ public void pageableQueryTest() {
...
+		this.userRepository.saveAll(users).blockLast();
+
+		List<String> pagedUsers = this.userRepository.findByAgeIn(Arrays.asList(22, 23, 24))
+				.map(User::getName)
+				.collectList()
+				.block();
+
+		assertThat(pagedUsers).containsExactly("Cloud");
+
+		pagedUsers = this.userRepository.findByAgeIn(Arrays.asList(17, 22))
```
Comment (elefeint) [inline spring-cloud-gcp-data-firestore/src/test/java/org/springframework/cloud/gcp/data/firestore/it/FirestoreRepositoryIntegrationTests.java]: Does not need to be for this PR, but using StepVerifier makes it easier to consume and verify a stream in one sentence. Then you don't need to block or save an intermediate variable, can just do `.expectNext(...).verifyComplete()`.

File: docs/src/main/asciidoc/firestore.adoc [hunk 9]
```diff
@@ -60,6 +60,22 @@ The Spring Boot starter for Google Cloud Firestore provides the following config
 |=========================================================
 
 
+==== Supported types
+Firestore  object conversion supports the following types:
```
Comment (meltsufin) [inline docs/src/main/asciidoc/firestore.adoc]: No `Date`??
```suggestion
You may use the following field types when defining your persistent entities or when binding query parameters:
```

File: docs/src/main/asciidoc/firestore.adoc [hunk 10]
```diff
@@ -60,6 +60,22 @@ The Spring Boot starter for Google Cloud Firestore provides the following config
...
+
+* Long
+* Integer
```
Comment (meltsufin) [inline docs/src/main/asciidoc/firestore.adoc]:
```suggestion
* `Integer`
```
Same formatting for others please.

File: docs/src/main/asciidoc/firestore.adoc [hunk 11]
```diff
@@ -60,6 +60,22 @@ The Spring Boot starter for Google Cloud Firestore provides the following config
...
+* Double
+* Float
+* String
+* Boolean
+* Character
+* Map
+* List
+* Enum
+
+POJOs containing fields of these types are also supported.
```
Comment (meltsufin) [inline docs/src/main/asciidoc/firestore.adoc]: Move this sentence above the list please to have it all in one sentence, which I already kind of did above.

**Pull request merged**